src/rate_limiter.rs

```diff
@@ -813,4 +812,32 @@ mod tests {
             _ => panic!("unexpected result"),
         }
     }
+
+    /// Checks that the rate-limiter's `overcharged` container can't be over-filled.
+    ///
+    /// Keeps trying to add GET requests for `ImmutableData` with a short delay between each
+    /// attempt. Most will fail, but this should ensure the `RateLimiter` always has an entry for
+    /// this client in its `used` container (in case we go back to using the absence of a client as
+    /// a trigger to purge their overcharged entries).
+    ///
+    /// After `OVERCHARGED_TIMEOUT_SECS` plus one minute has elapsed, there should not be an
+    /// excessive number of entries in the `overcharged` container.
+    #[test]
+    fn overcharged_limit() {
+        let mut rate_limiter = RateLimiter::new(false);
+        let client = IpAddr::from([0, 0, 0, 0]);
+        let wait_millis = MAX_IMMUTABLE_DATA_SIZE_IN_BYTES * 100 / RATE as u64;
+        let max_overcharged_entries = OVERCHARGED_TIMEOUT_SECS * RATE as u64 /
+            MAX_IMMUTABLE_DATA_SIZE_IN_BYTES;
+        let finish_time = FakeClock::now() + Duration::from_secs(OVERCHARGED_TIMEOUT_SECS + 60);
+        while FakeClock::now() < finish_time {
+            let name = XorName([0; XOR_NAME_LEN]);
+            let msg_id = MessageId::new();
+            let request = UserMessage::Request(Request::GetIData { name, msg_id });
+            let request_parts = unwrap!(request.to_parts(0));
+            let _ = add_user_msg_part(&mut rate_limiter, &client, &request_parts[0]);
+            FakeClock::advance_time(wait_millis);
+        }
+        assert!((rate_limiter.overcharged.len() as u64) <= max_overcharged_entries);
```

Comment:
@maqi:

I think your first calculation for `num_of_entries` is off by a factor of 1000 because `wait_millis` wasn't converted into seconds. Here's my working:

```
m = MAX_IMMUTABLE_DATA_SIZE_IN_BYTES
o = OVERCHARGED_TIMEOUT_SECS
r = RATE
w = wait_millis
```

Then the number of calls to `add_user_message_part` is:

```
n = 1000 * (o + 60) / w
n = 1000 * (o + 60) / (100m / r)
n = 10 * r * (o + 60) / m
```

So we have `n < max_overcharged_entries`, because `max_overcharged_entries = o * r / m`.